sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -274,22 +274,75 @@ case class And(left: Expression, right: Expression) extends BinaryOperator with
     val eval2 = right.gen(ctx)
 
     // The result should be `false`, if any of them is `false` whenever the other is null or not.
-    s"""
-      ${eval1.code}
-      boolean ${ev.isNull} = false;
-      boolean ${ev.value} = false;
+    if (!left.nullable && !right.nullable) {
+      ev.isNull = "false"
+      s"""
+        ${eval1.code}
+        boolean ${ev.value} = false;
 
-      if (!${eval1.isNull} && !${eval1.value}) {
+        if (${eval1.value}) {
+          ${eval2.code}
+          if (${eval2.value}) {
+            ${ev.value} = true;
+          }
+        }
+      """
+    } else {
+      if (!left.nullable) {
```

Comment:
The second and third branch do not look better than the fourth one (also no performance difference), could you remove that?